File: backfill_elo_from_csv.py
#!/usr/bin/env python3
"""
Recompute Elo from all matches in CSV (chronological order) and write SQL
to backfill matches.elo_change1/2 and players.elo_rating.

Usage: python3 backfill_elo_from_csv.py [path-to-csv]
Default CSV: matches_rows.csv in current dir.
Output: backfill-elo.sql
"""

import csv
import sys
from pathlib import Path

ELO_K = 32
ELO_INITIAL = 1000


def elo_expected(rating_a, rating_b):
    return 1 / (1 + 10 ** ((rating_b - rating_a) / 400))


def update_elo(r1, r2, player1_won):
    e1 = elo_expected(r1, r2)
    e2 = elo_expected(r2, r1)
    s1 = 1 if player1_won else 0
    s2 = 0 if player1_won else 1
    new1 = round(r1 + ELO_K * (s1 - e1))
    new2 = round(r2 + ELO_K * (s2 - e2))
    return new1, new2


def main():
    csv_path = sys.argv[1] if len(sys.argv) > 1 else Path(__file__).parent / "matches_rows.csv"
    rows = []
    with open(csv_path, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            if not row.get("id") or not row.get("created_at"):
                continue
            row["score1"] = int(row.get("score1", 0))
            row["score2"] = int(row.get("score2", 0))
            rows.append(row)

    rows.sort(key=lambda m: m["created_at"])
    ratings = {}
    match_updates = []

    for m in rows:
        p1 = m["player1_name"]
        p2 = m["player2_name"]
        r1 = ratings.get(p1, ELO_INITIAL)
        r2 = ratings.get(p2, ELO_INITIAL)
        player1_won = m["score1"] > m["score2"]
        new1, new2 = update_elo(r1, r2, player1_won)
        ratings[p1] = new1
        ratings[p2] = new2
        match_updates.append({
            "id": m["id"],
            "elo_change1": new1 - r1,
            "elo_change2": new2 - r2,
        })

    out_path = Path(__file__).parent / "backfill-elo.sql"
    lines = [
        "-- Backfill Elo from chronological match history. Run in Supabase SQL Editor.",
        "-- Matches (elo_change1, elo_change2):",
    ]
    for u in match_updates:
        lines.append(f"UPDATE public.matches SET elo_change1 = {u['elo_change1']}, elo_change2 = {u['elo_change2']} WHERE id = {u['id']};")
    lines.append("")
    lines.append("-- Players (final elo_rating):")
    for name, elo in sorted(ratings.items(), key=lambda x: -x[1]):
        safe_name = name.replace("'", "''")
        lines.append(f"UPDATE public.players SET elo_rating = {elo} WHERE name = '{safe_name}';")

    out_path.write_text("\n".join(lines), encoding="utf-8")
    print(f"Processed {len(rows)} matches, {len(ratings)} players.")
    print(f"SQL written to {out_path}")
    print("Final Elo:", sorted(ratings.items(), key=lambda x: -x[1]))


if __name__ == "__main__":
    main()

File: backfill-elo-from-csv.js
#!/usr/bin/env node
/**
 * Recompute Elo from all matches in CSV (chronological order) and output SQL
 * to backfill matches.elo_change1/2 and players.elo_rating.
 *
 * Usage: node backfill-elo-from-csv.js [path-to-csv]
 * Default CSV: ./matches_rows.csv (or pass path, e.g. "/Users/.../matches_rows (1).csv")
 * Output: backfill-elo.sql
 */

const fs = require("fs");
const path = require("path");

const ELO_K = 32;
const ELO_INITIAL = 1000;

function eloExpected(ratingA, ratingB) {
  return 1 / (1 + Math.pow(10, (ratingB - ratingA) / 400));
}

function updateElo(r1, r2, player1Won) {
  const E1 = eloExpected(r1, r2);
  const E2 = eloExpected(r2, r1);
  const S1 = player1Won ? 1 : 0;
  const S2 = player1Won ? 0 : 1;
  return {
    new1: Math.round(r1 + ELO_K * (S1 - E1)),
    new2: Math.round(r2 + ELO_K * (S2 - E2)),
  };
}

function parseCsv(content) {
  const lines = content.trim().split("\n");
  if (lines.length < 2) return [];
  const header = lines[0].split(",").map((h) => h.trim());
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(",").map((v) => v.trim());
    const row = {};
    header.forEach((h, j) => {
      row[h] = values[j] ?? "";
    });
    rows.push(row);
  }
  return rows;
}

const csvPath = process.argv[2] || path.join(__dirname, "matches_rows.csv");
const raw = fs.readFileSync(csvPath, "utf8");
const rows = parseCsv(raw);

const sorted = rows
  .filter((r) => r.id && r.player1_name && r.player2_name && r.created_at)
  .sort((a, b) => (a.created_at < b.created_at ? -1 : a.created_at > b.created_at ? 1 : 0));

const ratings = {};
const matchUpdates = [];

for (const m of sorted) {
  const p1 = m.player1_name;
  const p2 = m.player2_name;
  const r1 = ratings[p1] ?? ELO_INITIAL;
  const r2 = ratings[p2] ?? ELO_INITIAL;
  const player1Won = Number(m.score1) > Number(m.score2);
  const { new1, new2 } = updateElo(r1, r2, player1Won);
  ratings[p1] = new1;
  ratings[p2] = new2;
  matchUpdates.push({
    id: m.id,
    elo_change1: new1 - r1,
    elo_change2: new2 - r2,
  });
}

const outPath = path.join(__dirname, "backfill-elo.sql");
const sql = [
  "-- Backfill Elo from chronological match history. Run in Supabase SQL Editor.",
  "-- Matches (elo_change1, elo_change2):",
  ...matchUpdates.map(
    (u) =>
      `UPDATE public.matches SET elo_change1 = ${u.elo_change1}, elo_change2 = ${u.elo_change2} WHERE id = ${u.id};`
  ),
  "",
  "-- Players (final elo_rating):",
  ...Object.entries(ratings).map(
    ([name, elo]) =>
      `UPDATE public.players SET elo_rating = ${elo} WHERE name = '${name.replace(/'/g, "''")}';`
  ),
].join("\n");

fs.writeFileSync(outPath, sql, "utf8");

console.log(`Processed ${sorted.length} matches, ${Object.keys(ratings).length} players.`);
console.log(`SQL written to ${outPath}`);
console.log("Final Elo:", Object.entries(ratings).sort((a, b) => b[1] - a[1]));
